src/ifcopenshell-python/ifcopenshell/ids.py

```diff
@@ -76,17 +79,18 @@ class entity(facet):
     The IDS entity facet currently *with* inheritance
     """
 
-    parameters = ["name"]
-    message = "an entity name '%(name)s'"
+    parameters = ["name", "predefinedtype"]
     
-    # @todo predefinedtype
-
     def __call__(self, inst, logger):
-        logger.debug("Testing %s == %s", inst.is_a(), self.name)
         # @nb with inheritance
-        # return inst.is_a() == self.name
-        return facet_evaluation(inst.is_a(self.name), self.message % {"name": inst.is_a()})
-
+        if self.predefinedtype and hasattr(inst, "PredefinedType"):
```

Comment:
* I guess if `not hasattr(inst, "PredefinedType")` it means the IDS is invalid (for this particular IFC schema at least) and it should be reported somehow. The question is if we do a IDS validity check before running the model compliance check. ifcopenshell also has the schema data, so you can check whether an IDS is valid without actually opening a model

* PredefinedType can also be defined on the IfcTypeObject entity. (IsDefinedBy/RelatingType, IsTypedBy/RelatingType)

* In the IFC4.3 schema the PredefinedType attribute can be a bit more complex. It can be a SELECT (union type) of several ENUMs. In that case the attribute is handled slightly different in IfcOpenShell. At some point let's check if there is an example https://github.com/buildingSMART/Sample-Test-Files/tree/master/IFC%204.3

Reply:
from the IDS XSD I understood that IDS doesn't neceserilly need to have PredefinedType in entity:
`<xs:element name="predefinedtype" type="xs:anyType" minOccurs="0">` 
so it would still be a valid IDS file, right?

On the other hand, if IDS requires a particular PredefinedType and IFC (inst) has none, it should fail the validation, currently, it skips the validation. What do you think?

Reply:
> it should fail the validation, currently, it skips the validation. What do you think?

There's a subtle but important distinction here: (a) The attribute can be absent on the entity class or (b) the attribute can be unset (null, $).

I was hinting at these possibilities, like:

* A predefinedtype that isn't possible for an entity: like IfcWall SKYLIGHT
* A predefinedtype in an IDS for the IFC2X3 schema where the entity doesn't have a PredefinedType attribute.

In those cases I'd say the IDS itself is invalid.

It's also possible that the PredefinedType is unset or has value NOTDEFINED / USERDEFINED. I think those cases are properly handled (USERDEFINED is tricky perhaps) because you're simply comparing the value (can be `None`) to the value in the IDS XML.